src/main/scala/uk/gov/nationalarchives/omega/api/services/ApiService.scala

```diff
@@ -0,0 +1,219 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.gov.nationalarchives.omega.api.services
+
+import cats.effect.std.Queue
+import cats.effect.{ ExitCode, IO }
+import cats.implicits.catsSyntaxParallelTraverse_
+import jms4s.JmsAcknowledgerConsumer.AckAction
+import jms4s.config.QueueName
+import jms4s.jms.JmsMessage
+import jms4s.{ JmsAcknowledgerConsumer, JmsProducer }
+import org.typelevel.log4cats.slf4j.Slf4jFactory
+import org.typelevel.log4cats.{ LoggerFactory, SelfAwareStructuredLogger }
+import pureconfig.generic.auto._
+import pureconfig.{ ConfigObjectSource, ConfigSource }
+import uk.gov.nationalarchives.omega.api.common.ServiceArgs
+import uk.gov.nationalarchives.omega.api.conf.ServiceConfig
+import uk.gov.nationalarchives.omega.api.connectors.JmsConnector
+import uk.gov.nationalarchives.omega.api.services.LocalMessageStore.PersistentMessageId
+import uk.gov.nationalarchives.omega.api.services.ServiceState.{ Started, Starting, Stopped, Stopping }
+
+import java.nio.file.{ Files, Paths }
+import java.util.concurrent.atomic.AtomicReference
+
+class ApiService(val config: ServiceConfig) {
+
+  implicit val loggerFactory: LoggerFactory[IO] = Slf4jFactory[IO]
+  implicit val logger: SelfAwareStructuredLogger[IO] = LoggerFactory[IO].getLogger
+
+  private var errorOnInvalidServiceState = false
+  private val state = new AtomicReference[ServiceState](Stopped)
+
+  private val localMessageStore = {
+    val messageStoreFolder = Paths.get(config.tempMessageDir)
+    Files.createDirectories(messageStoreFolder)
+    new LocalMessageStore(messageStoreFolder)
+  }
+
+  def start: IO[ExitCode] = {
+    // check if we can start the service
+    if (!switchState(Stopped, Starting)(config.errorOnInvalidServiceState)) {
+      return IO.pure(ExitCode.Error)
+    }
+
+    // copy config option to local state
+    this.errorOnInvalidServiceState = config.errorOnInvalidServiceState
+
+    try {
+      // attempt startup
+      val ec = doStart(config)
+      // switch to the started state
+      switchState(Starting, Started)(errorOnInvalidServiceState)
+      ec
+    } catch {
+      case e: Throwable =>
+        // we are in some inconsistent state as startup failed, so we must attempt shutdown
+        try {
+          doStop()
+          switchState(Starting, Stopped)(errorOnInvalidServiceState)
+          IO.pure(ExitCode.Error)
+        } finally
+          println(e.getMessage)
+    }
+
+  }
+
+  @throws[IllegalStateException]
+  def stop(): Unit = {
+    // check if we can stop the service
+    if (!switchState(Started, Stopping)(errorOnInvalidServiceState)) {
+      return
```

Comment:
There's a return here, this could be re-written not to use it

Reply:
Refactored.